### ISSUE TITLE
TypedDict Literal Unification Failure in List Concatenation Return Expressions

### ISSUE THREAD

**Issue by CharString** [microsoft/pyright#9939]
**Describe the bug**

Maybe I misunderstand the PEP, but why does this unify on assignment, but not on return?

```python
class T(TypedDict):
    n: int


def c() -> list[T]:
    a: T = {"n": 1}
    b: T = {"n": 2}
    return [a] + ([b] if True else [])


def d() -> list[T]:
    return [{"n": 1}] + ([{"n": 2}] if True else [])


def e() -> list[T]:
    x: list[T] = [{"n": 1}]
    y: list[T] = [{"n": 2}] if True else []
    return x + y

def f() -> list[T]:
    z: list[T] = [{"n": 1}]
    return z + [{"n": 2}] if True else []

def g() -> T:
    return {"n":1}
```

`d` fails to unify, the rest does.

Since lists under concatenation (with [] as the identity) form a monoid, the type system should be able to “close” over that operation. In other words, if you expect a list[T] and you're building that list using concatenation, the operands should be contextually typed to T just as they are when assigned to a variable with an explicit type annotation. The fact that this works in assignments but fails in a direct return expression indicates that the expected type isn’t being propagated into the binary operator’s operands as it should be.

**Comment by erictraut** [comment]
Pyright is working as intended here, so I don't consider this a bug.

Binary operators like `+` in Python at runtime results in a series of calls to individual magic methods. For example, in the case of the `+` operator, the runtime will first call the `__add__` method on the first operand and pass the second operand as an argument. If that fails, it will then call the `__radd__` method on the second operand and pass the first operand as an argument. Static type checkers like pyright attempt to model this runtime logic.

To determine statically whether `__add__` or `__radd__` will be used, a type checker must first evaluate the types of the both operands. That means means the operand subexpressions must be evaluated without the benefit of [bidirectional type inference](https://microsoft.github.io/pyright/#/type-inference?id=bidirectional-type-inference-expected-types). That is, both subexpressions must be evaluated using default type evaluation rules without an "expected type context". That means that the subexpression `[{"n": 1}]` will evaluate to `list[dict[str, int]]` rather than `list[T]` in your example.

As you've shown in your code sample, there are ways to write the code that helps a static type checker understand your intent. You can use any of these techniques to make your code type check without error.
